File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/schema/GQLRootQueryValidationRuns.java
package net.nemerosa.ontrack.graphql.schema;

import graphql.schema.DataFetcher;
import graphql.schema.GraphQLFieldDefinition;
import graphql.schema.GraphQLNonNull;
import net.nemerosa.ontrack.model.structure.ID;
import net.nemerosa.ontrack.model.structure.StructureService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Collections;

import static graphql.Scalars.GraphQLInt;
import static graphql.schema.GraphQLArgument.newArgument;
import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;
import static net.nemerosa.ontrack.graphql.support.GraphqlUtils.stdList;

@Component
public class GQLRootQueryValidationRuns implements GQLRootQuery {

    private final StructureService structureService;
    private final GQLTypeValidationRun validationRun;

    @Autowired
    public GQLRootQueryValidationRuns(StructureService structureService, GQLTypeValidationRun validationRun) {
        this.structureService = structureService;
        this.validationRun = validationRun;
    }

    @Override
    public GraphQLFieldDefinition getFieldDefinition() {
        return newFieldDefinition()
                .name("validationRuns")
                .type(stdList(validationRun.getType()))
                .argument(
                        newArgument()
                                .name("id")
                                .description("ID of the validation run to look for")
                                .type(new GraphQLNonNull(GraphQLInt))
                                .build()
                )
                .dataFetcher(validationRunFetcher())
                .build();
    }

    private DataFetcher validationRunFetcher() {
        return environment -> {
            Integer id = environment.getArgument("id");
            if (id != null) {
                // Fetch by ID
                return Collections.singletonList(
                        structureService.getValidationRun(ID.of(id))
                );
            }
            // Empty list
            else {
                return Collections.emptyList();
            }
        };
    }

}

File: ontrack-ui-graphql/src/main/java/net/nemerosa/ontrack/graphql/schema/GQLRootQueryValidationRuns.kt
package net.nemerosa.ontrack.graphql.schema

import graphql.Scalars.GraphQLInt
import graphql.schema.DataFetcher
import graphql.schema.GraphQLFieldDefinition
import graphql.schema.GraphQLFieldDefinition.newFieldDefinition
import graphql.schema.GraphQLNonNull
import net.nemerosa.ontrack.graphql.support.GraphqlUtils.stdList
import net.nemerosa.ontrack.model.structure.ID
import net.nemerosa.ontrack.model.structure.StructureService
import net.nemerosa.ontrack.model.structure.ValidationRun
import org.springframework.stereotype.Component

@Component
class GQLRootQueryValidationRuns(
        private val structureService: StructureService,
        private val validationRun: GQLTypeValidationRun
) : GQLRootQuery {

    override fun getFieldDefinition(): GraphQLFieldDefinition {
        return newFieldDefinition()
                .name("validationRuns")
                .type(stdList(validationRun.typeRef))
                .argument {
                    it.name("id")
                            .description("ID of the validation run to look for")
                            .type(GraphQLNonNull(GraphQLInt))
                }
                .dataFetcher(validationRunFetcher())
                .build()
    }

    private fun validationRunFetcher(): DataFetcher<List<ValidationRun>> {
        return DataFetcher { environment ->
            val id: Int? = environment.getArgument("id")
            if (id != null) {
                // Fetch by ID
                listOf(structureService.getValidationRun(ID.of(id)))
            } else {
                emptyList()
            }
        }
    }

}
